ts41/get.d.ts

```diff
@@ -0,0 +1,123 @@
+import {Split} from './utilities';
+import {StringDigit} from '../source/utilities';
+
+/**
+Like @see Get but receives an array of strings as a path parameter.
+*/
+type GetWithPath<ObjectType, Keys extends readonly string[]> =
+	Keys extends []
+	? ObjectType
+	: Keys extends [infer Head, ...infer Tail]
+	? GetWithPath<PropertyOf<ObjectType, Extract<Head, string>>, Extract<Tail, string[]>>
+	: never;
+
+/**
+Splits a dot-prop style path into a tuple comprised of the properties in the path. Handles square-bracket notation.
+
+@example
+```
+ToPath<'foo.bar.baz'>
+//=> ['foo', 'bar', 'baz']
+
+ToPath<'foo[0].bar.baz'>
+//=> ['foo', '0', 'bar', 'baz']
+```
+*/
+type ToPath<S extends string> = Split<FixPathSquareBrackets<S>, '.'>;
+
+/**
+Replaces square-bracketed dot notation with dots, for example, `foo[0].bar` -> `foo.0.bar`.
+*/
+type FixPathSquareBrackets<Path extends string> =
+	Path extends `${infer Head}[${infer Middle}]${infer Tail}`
+	? `${Head}.${Middle}${FixPathSquareBrackets<Tail>}`
+	: Path;
+
+/**
+Returns true if `S` is made up out of `C` repeated 0 or more times.
+
+@example
+```
+ConsistsOnlyOf<'aaa', 'a'> //=> true
+ConsistsOnlyOf<'aBa', 'a'> //=> false
+ConsistsOnlyOf<'', 'a'> //=> true
+```
+*/
+type ConsistsOnlyOf<LongString extends string, Substring extends string> =
+	LongString extends ''
+	? true
+	: LongString extends `${Substring}${infer Tail}`
+	? ConsistsOnlyOf<Tail, Substring>
+  : false;
+
+/**
+Convert a type which may have number keys to one with string keys, making it possible to index using strings retrieved from template types.
+
+@example
+```
+type WithNumbers = {foo: string; 0: boolean};
+type WithStrings = WithStringKeys<WithNumbers>;
+
+type WithNumbersKeys = keyof WithNumbers;
+//=> 'foo' | 0
+type WithStringsKeys = keyof WithStrings;
+//=> 'foo' | '0'
+```
+*/
+type WithStringKeys<ObjectType extends Record<string | number, any>> = {
+	[Key in `${Extract<keyof ObjectType, string | number>}`]: ObjectType[Key]
+};
+
+/**
+Get a property of an object or array. Works when indexing arrays using number-literal-strings, for example, `PropertyOf<number[], '0'> = number`, and when indexing objects with number keys.
+
+Returns `unknown` if `Key` is not a property of `ObjectType`.
+*/
+type PropertyOf<ObjectType, Key extends string> =
+	ObjectType extends null | undefined
+	? undefined
+	: Key extends keyof ObjectType
+	? ObjectType[Key]
+	: ObjectType extends ArrayLike<infer Item>
```

Comment:
If `Key` is eg. `'length'` then `Key extends keyof ObjectType` will be truthy if `ObjectType` is an array.

I think we therefore need to handle the array case first to avoid this?

And in regards to the array case, a couple of things:

First:

`ArrayLike<>` is barely documented at all, neither in this repo nor in the official TypeScript docs, I had to go to the source code of TypeScript itself to find what it does: https://github.com/microsoft/TypeScript/blob/d0c28ab0c2c54c8d4ab2f0c34385aed501ef73c6/src/lib/es5.d.ts#L1431-L1434

Are there any benefits of this over `(infer Item)[]`, `Array<infer Item>` or even `{ [n: number]: infer Item; }`?

Secondly:

I would suggest splitting the logic for arrays out of this + ensure that it works correctly with array literals, like this:

```typescript
// Safe to assume that all array literals have a 0 key?
type isArrayLiteral<ArrayType> = '0' extends keyof ArrayType ? true : false;

type KeyOfArray<ArrayType extends any[], Key extends string> =
  // One should only use numeric types for arrays, so any other keys gets the "never" treatment
  ConsistsOnlyOf<Key, StringDigit> extends false
  ? never
  // If we have found ourselves an array literal, then get value by key
  : isArrayLiteral<ArrayType> extends true
  ? ArrayType[Key]
  // Else infer type
  : ArrayType extends (infer ItemType)[]
  ? ItemType
  // This won't ever happen
  : never
)
```

Using that we could rewrite this part of `PropertyOf` to something like:

```suggestion
	: ObjectType extends any[]
	? KeyOfArray<ObjectType, Key>
	: Key extends keyof ObjectType
	? ObjectType[Key]
```

Reply:
I'll try this out - I agree, `ArrayLike` is weird. FWIW I _think_ we'd be ok if the key is `length` as the other check would just get `ObjectType['length']` which would resolve to `number`.

Also I'm not sure it is safe to assume all array/tuple literals have a '0' key. `Parameters<() => {}>` is a not-totally-unrealistic counterexample which has type `[]`.

Reply:
Ok, here are the various tradeoffs to each approach:

`Array<infer Item>` of `(infer Item)[]` won't work without a special case for `ReadonlyArray`. Failing test:
```typescript
type Test = { a: ReadonlyArray<{b: 1 }> }
expectTypeOf<Get<Test, 'a[0].b'>>().toEqualTypeOf<1>()
```

`{ [n: number]: infer Item }` breaks number-indexed types. This test passes when it shouldn't:
```typescript
type Test = { 1: { b: string } }
expectTypeOf<Get<Test, '123456.b'>>().toBeString()
```

The following all work:
1. `ReadonlyArray<infer Item>`
2. `ArrayLike<infer Item>`
3. `{ length: number; [n: number]: infer Item }` (equivalent to ArrayLike)

Given this conversation, I'd like to go for 3 to avoid using a little-documented type like in 2, or relying on `Array` being assignable to `ReadonlyArray` like in 1.

Tests for all these cases already exist in `test-d/get.ts` - feel free to try them out and let me know if you find a better fit. (Note: I had to un-exclude `"test-d/**/*"` from tsconfig to see errors in my IDE).